### DISCUSSION TITLE
Error attempting to optimize with direction holes.

### DISCUSSION THREAD

**Discussion by JeffBurris** [edwardkort/WWIDesigner#124] [NAF Study] [answered]
Hi Mr. Kort and company!

**Update**:
I'm editing this to amend my issue to:
"WID stalls or does not produce an optimized instrument file (when attempting to incorporate direction holes)".
Before, I focused  on "the error report" because I only now noticed I get a similar-looking "error" when WID works fine for me and does produce an optimized instrument file okay (when not attempting to incorporate direction holes).
I will post this other "error" it yields when things are okay in an ammendment below (to compare)since it is a helpful activity report anyway...

First, here's my original post:

 - - - - - - 

I've been doing just peachy keen with WID for NAF Studies until trying this and got a new error about which I'm puzzled. 
The issue is probably user error on my part, as I see nothing on the Issues page.
Optimization etc. has worked great for me, producing numerous, physical flutes matching WID on the nose.

I watched the video on doing direction holes but, instead of using your exact figures, I gave my first whirl different values that match what I'm usually doing. I wanted to add direction holes (and length) to essentially my same A4 flute that has been working out fine from A-Z, prior.

I used the wizards for the new Tuning File, as well as a Constraints File, per the tutorial but merely entered slightly different specs (copied over from my favorite no-direction-holes A4 flute I did with WID). I used the same specs as you for the direction holes.

What I get is:
(begin copy-paste)
 - - - - - - - - - - - - - - - - - 
System has 15 optimization variables and 14 target notes.
Initial error: 179197.20208786847.
Exception: null
java.util.NoSuchElementException
	at java.util.LinkedList$ListItr.next(Unknown Source)
	at com.wwidesigner.geometry.Instrument.processPosition(Instrument.java:499)
	at com.wwidesigner.geometry.Instrument.updateComponents(Instrument.java:417)
	at com.wwidesigner.optimization.HoleGroupPositionFromTopObjectiveFunction.setGeometryPoint(HoleGroupPositionFromTopObjectiveFunction.java:162)
	at com.wwidesigner.optimization.MergedObjectiveFunction.setGeometryPoint(MergedObjectiveFunction.java:103)
	at com.wwidesigner.optimization.BaseObjectiveFunction.getErrorVector(BaseObjectiveFunction.java:200)
	at com.wwidesigner.optimization.BaseObjectiveFunction.value(BaseObjectiveFunction.java:125)
	at org.apache.commons.math3.optim.nonlinear.scalar.MultivariateOptimizer.computeObjectiveValue(MultivariateOptimizer.java:115)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.bobyqb(BOBYQAOptimizer.java:835)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.bobyqa(BOBYQAOptimizer.java:340)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.doOptimize(BOBYQAOptimizer.java:252)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.doOptimize(BOBYQAOptimizer.java:49)
	at org.apache.commons.math3.optim.BaseOptimizer.optimize(BaseOptimizer.java:153)
	at org.apache.commons.math3.optim.BaseMultivariateOptimizer.optimize(BaseMultivariateOptimizer.java:65)
	at org.apache.commons.math3.optim.nonlinear.scalar.MultivariateOptimizer.optimize(MultivariateOptimizer.java:63)
	at com.wwidesigner.optimization.ObjectiveFunctionOptimizer.runBobyqa(ObjectiveFunctionOptimizer.java:515)
	at com.wwidesigner.optimization.ObjectiveFunctionOptimizer.runBobyqa(ObjectiveFunctionOptimizer.java:534)
	at com.wwidesigner.optimization.ObjectiveFunctionOptimizer.optimizeObjectiveFunction(ObjectiveFunctionOptimizer.java:225)
	at com.wwidesigner.gui.StudyModel.optimizeInstrument(StudyModel.java:948)
	at com.wwidesigner.gui.StudyView.optimizeInstrument(StudyView.java:392)
	at com.wwidesigner.gui.WIDesigner$16.activityPerformed(WIDesigner.java:715)
	at com.jidesoft.app.framework.activity.Activity.b(Unknown Source)
	at com.jidesoft.app.framework.activity.Activity$1.run(Unknown Source)
	at java.util.TimerThread.mainLoop(Unknown Source)
	at java.util.TimerThread.run(Unknown Source)

 - - - - - - - - - - - - - - - 
_end copy/paste_ 
(eternal pause and no optimized instrument file)
* It stops there and never reports anything else, like it does in a successful scenario without direction holes *

**  If you'd like to see them, I have attached Instrument, Constraints, and Tuning files... in order for GidHub forum to accept them I have changed the file extensions temporarily to .txt instead of .xml. They are named accordingly.

[constraints-grouped-HolePosition-andSize-A4-withDirHoles.txt](https://github.com/edwardkort/WWIDesigner/files/6208795/constraints-grouped-HolePosition-andSize-A4-withDirHoles.txt)
[Tuning-file-A4-w-DirectionHoles.txt](https://github.com/edwardkort/WWIDesigner/files/6208798/Tuning-file-A4-w-DirectionHoles.txt)
[instrument-file-A4-w-direction-holes-BoreLengthCorrected.txt](https://github.com/edwardkort/WWIDesigner/files/6209048/instrument-file-A4-w-direction-holes-BoreLengthCorrected.txt)

- - - - -
**Update**: Later I went back to make sure my no-direction-holes process was still working fine, and it was.
Only then, though, did I notice I get a similar "error" and report even though no-direction-holes optimization **produces an optimized instrument file just fine:**
(begin copy/paste)
 - - - - - - - - - - - - - - - - - -
 System has 15 optimization variables and 14 target notes.
Initial error: 12765.560556034605.
Exception: null
java.util.NoSuchElementException
	at java.util.LinkedList$ListItr.next(Unknown Source)
	at com.wwidesigner.geometry.Instrument.processPosition(Instrument.java:499)
	at com.wwidesigner.geometry.Instrument.updateComponents(Instrument.java:417)
	at com.wwidesigner.optimization.HoleGroupPositionFromTopObjectiveFunction.setGeometryPoint(HoleGroupPositionFromTopObjectiveFunction.java:162)
	at com.wwidesigner.optimization.MergedObjectiveFunction.setGeometryPoint(MergedObjectiveFunction.java:103)
	at com.wwidesigner.optimization.BaseObjectiveFunction.getErrorVector(BaseObjectiveFunction.java:200)
	at com.wwidesigner.optimization.BaseObjectiveFunction.value(BaseObjectiveFunction.java:125)
	at org.apache.commons.math3.optim.nonlinear.scalar.MultivariateOptimizer.computeObjectiveValue(MultivariateOptimizer.java:115)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.bobyqb(BOBYQAOptimizer.java:835)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.bobyqa(BOBYQAOptimizer.java:340)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.doOptimize(BOBYQAOptimizer.java:252)
	at org.apache.commons.math3.optim.nonlinear.scalar.noderiv.BOBYQAOptimizer.doOptimize(BOBYQAOptimizer.java:49)
	at org.apache.commons.math3.optim.BaseOptimizer.optimize(BaseOptimizer.java:153)
	at org.apache.commons.math3.optim.BaseMultivariateOptimizer.optimize(BaseMultivariateOptimizer.java:65)
	at org.apache.commons.math3.optim.nonlinear.scalar.MultivariateOptimizer.optimize(MultivariateOptimizer.java:63)
	at com.wwidesigner.optimization.ObjectiveFunctionOptimizer.runBobyqa(ObjectiveFunctionOptimizer.java:515)
	at com.wwidesigner.optimization.ObjectiveFunctionOptimizer.runBobyqa(ObjectiveFunctionOptimizer.java:534)
	at com.wwidesigner.optimization.ObjectiveFunctionOptimizer.optimizeObjectiveFunction(ObjectiveFunctionOptimizer.java:225)
	at com.wwidesigner.gui.StudyModel.optimizeInstrument(StudyModel.java:948)
	at com.wwidesigner.gui.StudyView.optimizeInstrument(StudyView.java:392)
	at com.wwidesigner.gui.WIDesigner$16.activityPerformed(WIDesigner.java:715)
	at com.jidesoft.app.framework.activity.Activity.b(Unknown Source)
	at com.jidesoft.app.framework.activity.Activity$1.run(Unknown Source)
	at java.util.TimerThread.mainLoop(Unknown Source)
	at java.util.TimerThread.run(Unknown Source)

System has 11 optimization variables and 14 target notes.
Initial error: 22023.263823160625.
Performed 4740 tuning calculations in 316 error norm evaluations.
Final error:  1402.9130025494724.
Residual error ratio: 0.06370141200752033
Elapsed time: 0.4 seconds.
 - - - - - - - - - - - - - - - - - - - 
 (end copy/paste)

Then it completes and spits out a lovely, optimized instrument file.
As you can see, this latter report, when successfully yielding an optimized instrument file, finishes up with six more lines of confirmative text at the end, which the other, unsuccessful scenario never does.

I tried waiting an extra long tim, but nothing happens. 
It's a pretty up to date machine that ptherwise optimizes in seconds.

**As always, thank you for the priceless gifts you have given the community.**

I _sincerely appreciate_ any insight, thank you.
Jeff

**Comment by edwardkort** [comment, marked as answer]
Hi Jeff,

The stack trace you are seeing is due to the optimizer trying to find a solution where a particular trial bore length excludes either holes (in this case) or bore points. I try to trap such instances gracefully, but you have really given the optimizer an impossible problem.

Please review the Direction Hole video starting at about 16 minutes in. You will see that, for the problem you are trying to solve, a fixed bore length (equal lower and upper bounds) must be set in the constraints.

If, in your constraints file, I set the bore length to 15.125", WID quickly converges to a solution. It is not a very good solution, primarily due to the large (0.31) lower bound of the top-hole ratio. 

In general, as soon as you have a rough idea of the final bore length, you should refine the bore-length constraint.

And BTW, I received notification of your YouTube question concerning note frequency values, although it hasn't shown up on YouTube. Anyway, in answer to your question... Your "by-the-book" values are rounded to 2 decimal places, WID rounds them to 4 decimal places. The values in WID are more accurate but still not exact. Nonetheless, the difference in frequency based on 2 or 4 decimal place accuracy is irrelevant.

Cheers,
Edward Kort
Wood Wind

**Reply by JeffBurris** [reply]
Thanks so much Mr. Kort.
I was sure it was an oversight on my part. My apologies. I thank you very, very much for your time.

As far as the frequencies, thanks for addressing that as well. I'd erased the question since I knew it was insignificant in practice and my curiosity where they came from mightn't have been worth bothering you with, hehe.

I'll try to be more thorough going forward with the software. Thanks again for your help.
Sincerely
Jeff

**Reply by JeffBurris** [reply]
Hi Mr Kort, 
I started looking at things again and please note I am not in the least challenging your information; I'm just trying to understand enough to use as a point of departure. Thanks for bearing with me. This is so important to me and even others in my family.
We already owe you so much.

First, I will give it another go with studying the nodal interference chart-aided process of determining the best lower bound for the top-hole ratio. I thought when the square-ended A4  flute with 3/5" ID and walls about .188 thick was aboput 13 inches long it yielded a .295 for that top hole ratio spec-- does that still sound inordinately large? Then that top note was a little weak even with .26" min. size hole, so I thought maybe I was too close.
Some of the guys over at the maker's forum whose flutes I play and who helped me with WID have been telling me they go through that routine with the nodal interference chart (like you show in a video) and add a .01 to the result (evidently from experience or for physical error wiggle room), so with .295 essentially being .30 and with my adding a .01 I arrived at .31.
But you are saying that's very large.
Let me delve back into all that again and check my comprehension and work...
Jon looked over his ratio numbers for that Dimensionless spec and said he never enters lower than .29 on a mid-sized flute. I see you have a default of .25 but I've never gotten anything nearly that low.

I'm still trying to for a better, overall, well-rounded technical comprehension of the acoustics.
Then I wouldn't so easily make oversights from getting frustrated or for me to not know what something in a lesson or kind answer means.

I'm the odd man out in a family othwerwise made of engineers and pilots. I got by in the aerospace machine shops but never got a degree until I was 48 and I struggle with a dyslexic-like condition exascerbated by the fact people think it doesn't exist. 
I know for a fact I have to try harder than some to conquer feats such as understanding everything you or Mike Prairie say.

I will keep at it. It's always darkest before dawn. I've been struggling with some aspects of flute making since last August.

I decided on a 15.125 inch bore length instead of, say, 16 like in the video, because I'd already cut the flute blank before deciding for certain reasons to do direction holes. I don't have to use this blank... are you saying the 15.125" length is also a bit short for this? 

I don't mean to take up more of your time. I'll go back to some of the basics I was trying not to dwell too long in to see if I can fortify my overall, comprehensive grasp of it all. 

I will go see now what you considered a poor solution and see what I can do to enable WID and physics to yield a better one.

What I consider good so far is this same flute with a 13" bore without direction holes with a deviation chart like:

![devs](https://user-images.githubusercontent.com/75704396/112702328-804a5f00-8e69-11eb-9a1d-c7e2d5800e00.png)

...arrived upon by using tuning file weights of 1's, 2's, and 3's with one 6.
It was the process of much fiddling. I would be curious what you think of that deviation chart for a single-ID straight bore, quasi-chromatic NAF.

Thankyou for any and all time you ever spend helping me on this to which I remain dedicated,
In earnest,
Jeff

**Reply by edwardkort** [reply]
Hey Jeff,

If I can help, I'm glad to spend the time.

If I put your numbers (0.75" bore diameter, 13" bore length, exact A4 tuning) into Mike Prairie's spreadsheet, I get a top-hole/bore-length ratio of 0.2788. And here, accuracy counts. Did your bore diameter end up exactly 0.75", and is the bore length exactly 13"?

With my flutes, the top-hole ratio is typically between 0.24 and 0.26. I can get away with using those exact numbers in the final hole layout, but I typically add 0.01 to that ratio. The lower the flute, the less this ratio drives the hole layout. And I don't know whether Jon pushes to get the second-octave notes in tune. Without them in the calculation, you can use a quite large top-hole ratio successfully.

And sometimes you can be fooled into thinking you have nodal interference when the upper notes in the first octave sound weak. Those notes are much more sensitive to splitting edge position, up and down, than any of the other notes. Often just lowering the splitting edge a couple of thousandths of an inch will give you robust voicing.

The total bore length, when using direction holes, is up to you. However, it turns out that it does affect tuning and how in-tune the flute can be optimized. When I directed you to fix the bore length in the constraints, that is really a simplification. As long as the lower bound for the bore length is greater than the length of a flute with no direction holes, plus room for the direction holes, you can let the WID optimize bore length. Try setting the lower bound to 13.5" and the upper bound to 18" and run the optimizer. I was surprised with the result. But I almost never use direction holes.

The tuning of your no-direction-holes A4 flute is acceptable. My gold standard is less than 5 cents deviation for any note, but I jump through a lot of hoops to achieve this, both during the design and construction phases.

I think that you are doing fine. Keep at it.

Cheers,
Edward Kort
Wood Wind

**Reply by JeffBurris** [reply]
Fantastic Mr. Kort. 
This is the type of thing that helps a lot. 
Between this and the review, tonight is turning out to be a relieving step forward.

I aim for the "reedy, less pure" sound the way the harmonics seem to work out with a splitting edge around the middle of the air stream (50/50 between flue floor and ceiling), so I mill the flue (let's say it's .040" deep) from the SAC hole on through half the rough TSH hole , stop, and raise the spindle 1/2 way (let's say to .020" deep) so my splitting edge is automatically in that halfway spot. Now, however, I think I'll start raising it a bit more so I can ease it on down manually with the file as I test it, since you're confirming it can make such an important difference (such as a weaker top note) and isn't just subtle, "signature" voicing. I was looking everywhere else.

What's curious is that everyone seems to place the splitting edge anywhere from (a) halfway between the flue ceiling and floor to (b) all the way at the bottom (for pure sound?) but no one mentions putting it almost at the top of the air stream. I have a bamboo flute I made (the second one ever) where the splitting edge is so high it's virtually on the same plane upon which the block rests (edge filed infinitessimally down, only to make less blunt). It sounds wonderful (maybe a little hiss I suspected would lessen with little bevels on the TSH & block where the air exits the flue). 

Do you think the harmonics phenomenon (splitting edge producing a more pure sound toward the bottom) could be mirrored on either distant side of the half way mark, so that having a very high splitting edge mimicks a very low one, at least when it comes to this nuance? I'd do a comparison with the spectrum analyzer but my flutes have too many variables. I'd need to make everything else the same do do a side by side.

I believe I'm almost there when it comes to avoiding major rookie issues, so I can explore preferences and work toward expertise. I'm my own worst critic because there are acquaintances already wanting to buy when they examine my better prototype(s), but I want these to retain a good impression after folks compre with other flutes, or after their playing improves and they notice the second octave etc. 
I've had several play very nicely but I also need consistency and predictability.

With gratitude,
Jeff

**Reply by edwardkort** [reply]
Hi Jeff,

I prefer a very clean tone. For my flutes, any breathiness or high-harmonic chaff is unacceptable. Keep this in mind in my response below.

I use a flue depth of 0.032", with a tolerance of only about 0.001". Anything taller or shallower impacts the tone. I also aim for a TSH length of about 0.18", although my flutes tolerate a very large variation in this value. And I use a moveable splitting edge, so changing the splitting edge geometry and position are independent.

In the past, I would start with the splitting edge even with the top of the flue. Then during the tuning process, I would lower the splitting edge just enough to voice the upper notes. But I noticed that I always ended up with the splitting edge in the same position vertically, about centered in the flue. So now I just put it there from the start.

I have never found the tone with the splitting edge even with the bottom of the flue to be acceptable. So no, there is no vertical symmetry there.

My joy in making flutes comes from the realization of my musical intent in a physical instrument through control of aesthetics, tonality, tuning, and playing characteristics. If I couldn't take 6 pieces of wood and make 6 identical flutes,  I wouldn't stand a chance of controlling all the variables to make a flute reflect my intent. So consistency is really critical.

Hope this helps.

Cheers,
Edward

**Comment by JeffBurris** [comment]
Okay thanks!
I agree and want total predictability.
This is where I can benefit from my experience with a set or calipers and experience with very low tolerances of +/- .001".

It's curious that some very well-established makers disagree about pure vs. more rich with harmonics and the position  of the splitting edge, which resides at the bottom vs. the middle of the flue. Your experience aligns with Mr. Prairie's.

Do you agree that odd harmonics make something sound more rich, and the even harmonics make it sound more clean? If so, then the different testimonies surely lie in which position yields which, and not what constitutes clean or reedy tone. One would intuitively think odd harmonics to be more dissonant combined with the primary target frequency.
In any case It's simple enough for me to eventually get around to putting mine on a spectrum analysis and see what it is I'm liking the best.

I am torn between what I like and what most people might want​.
To me, what I'd consider the purest tone can sound a like my recorder or clean FM synthesis, whereas I prefer to produce just a bit more "primitive" or "traditional" sound (but not too much, esp. not too breathy). A little character stirs different emotions in me and makes me imagine it's a fabulous but rim-blown instrument.  Perhaps I could get the best "attenuated primitive" effect by combining a particular splitting edge position with a particular wall thickness (since thinner walls seem to yield some harmonics thicker ones mute).

But whether collectors with strong preferences would be would be divided equally, or if more prefer a very, very pure tone, I do not know.

Thanks for your thoughts, they have been helpful and enjoyed.
Jeff




________________________________

**Reply by edwardkort** [reply]
Jeff,

Each player has a preferred timbre for their flutes. Personally, I like the bell-like tone of even-number harmonics (the fundamental is H0). Operationally, I define mellowness by the H0/H1 ratio. I place the splitting edge to yield a clean tone. But I control the general balance of harmonics by the bore diameter. I spend a lot of time prototyping identical flutes differing only in their bore diameters. And other factors are also in play. I can hear a timbre difference (as can others) between a chimney bird end and a simple slant face; and I prefer to slant face.

If your market is custom flutes (my preference), then you can make a flute with those characteristics (tonal and otherwise) valued by that customer. If mass marketing (flute festivals, etc.), I recommend making the kind of flute that you like. You will then find your market niche.

Cheers,
Edward
